### PR TITLE
Add application of BQSR to HaplotypeCaller function

### PR DESCRIPTION
**Pull request by mp15**
Tests added

### CODE REVIEW

File: lib/npg_pipeline/function/haplotype_caller.pm [hunk 1]
```diff
@@ -94,8 +96,23 @@ sub create {
         $chuncking_base_name,
         $product->chunk;
       my $output_path = $product->file_path($out_dir_path, ext => 'g.vcf.gz');
-      my $command = sprintf q{%s %s %s -I %s -O %s -L %s},
-        $self->gatk_cmd, $GATK_TOOL_NAME, $gatk_args, $input_path, $output_path, $region;
+      my $command;
+
+      if ($self->bqsr_enable($product) && $self->bqsr_apply_enable($product)) {
```
Comment (mgcam) [inline lib/npg_pipeline/function/haplotype_caller.pm]: Do we need bqsr_apply_enable?

File: lib/npg_pipeline/function/haplotype_caller.pm [hunk 2]
```diff
@@ -94,8 +96,23 @@ sub create {
...
+        # TODO: write proper temporary file handling support
```
Comment (mgcam) [inline lib/npg_pipeline/function/haplotype_caller.pm]: temp dir in the product archive dir?

File: lib/npg_pipeline/function/haplotype_caller.pm [hunk 3]
```diff
@@ -94,8 +96,23 @@ sub create {
...
+        my $temp_path = $product->file_path($out_dir_path, ext => 'temp.cram');
+        my $gatk_cmd = sprintf q{%s %s %s -I %s -O %s -L %s},
+          $self->gatk_cmd, $GATK_TOOL_NAME, $gatk_args, $input_path, $temp_path, $region;
```
Comment (mgcam) [inline lib/npg_pipeline/function/haplotype_caller.pm]: Should $temp_path be input here and $output_path the output?

File: lib/npg_pipeline/function/haplotype_caller.pm [hunk 4]
```diff
@@ -94,8 +96,23 @@ sub create {
...
+
+        my $bqsr_args = "-R $ref_path --preserve-qscores-less-than 6 --static-quantized-quals 10 --static-quantized-quals 20 --static-quantized-quals 30";
+        my $bqsr_cmd = sprintf q(%s %s %s -I %s -O %s -L %s),
+          $self->gatk_cmd, $GATK_BQSR_TOOL_NAME, $bqsr_args, $temp_path, $output_path, $region;
```
Comment (mgcam) [inline lib/npg_pipeline/function/haplotype_caller.pm]: Should $temp_path be the output here and $input_path the input?

File: lib/npg_pipeline/function/haplotype_caller.pm [hunk 5]
```diff
@@ -94,8 +96,23 @@ sub create {
...
+        my $rm_cmd = "rm \"$temp_path\" \"$temp_path.crai\"";
+        $command = join ' && ', [$bqsr_cmd, $gatk_cmd, $rm_cmd];
```
Comment (mgcam) [inline lib/npg_pipeline/function/haplotype_caller.pm]: Do we want to have non-zero exit if the rm command fails?

File: lib/npg_pipeline/function/haplotype_caller.pm (hunk 1 above)
Comment (mp15) [reply lib/npg_pipeline/function/haplotype_caller.pm]: I am inclined to say yes. There are some projects out there producing the BQSR table but not applying it.

File: lib/npg_pipeline/function/haplotype_caller.pm (hunk 4 above)
Comment (mp15) [reply lib/npg_pipeline/function/haplotype_caller.pm]: Yes

File: lib/npg_pipeline/function/haplotype_caller.pm (hunk 3 above)
Comment (mp15) [reply lib/npg_pipeline/function/haplotype_caller.pm]: Yes

File: lib/npg_pipeline/function/haplotype_caller.pm (hunk 5 above)
Comment (mp15) [reply lib/npg_pipeline/function/haplotype_caller.pm]: Fixed

**Pull request merged**